### PR TITLE
Created IpHelper

### PR DESCRIPTION
**Pull request by SilverFire**
Replaces #14043

### CODE REVIEW

File: framework/helpers/BaseIpHelper.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
+<?php
+
+namespace yii\helpers;
+
+/**
+ * Class BaseIpHelper
+ *
+ * @author Dmytro Naumenko <d.naumenko.a@gmail.com>
```
Comment (dmirogin) [inline framework/helpers/BaseIpHelper.php]: since tag needed

File: framework/helpers/BaseIpHelper.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        $parts = explode('/', $range);
+        $net = array_shift($parts);
+        $range_cidr = array_shift($parts);
+
+        $netVersion = static::getIpVersion($net);
+        if ($ipVersion !== $netVersion) {
+            return false;
+        }
+        if ($range_cidr === null) {
+            $range_cidr = $netVersion === 4 ? self::IPV4_ADDRESS_LENGTH : self::IPV6_ADDRESS_LENGTH;
```
Comment (dmirogin) [inline framework/helpers/BaseIpHelper.php]:
```php
$netVersion === 4
```
4 -> self::IPV4

**Review by dmirogin — APPROVED** [review]

File: framework/validators/IpValidator.php [hunk 3]
```diff
@@ -354,14 +355,14 @@ private function validateSubnet($ip)
             return [$this->message, []];
         }
 
-        if ($this->getIpVersion($ip) == 6) {
+        if ($this->getIpVersion($ip) === IpHelper::IPV6) {
             if ($cidr !== null) {
-                if ($cidr > static::IPV6_ADDRESS_LENGTH || $cidr < 0) {
```
Comment (samdark) [inline framework/validators/IpValidator.php]: Should constants be deprecated?

**Review by samdark — APPROVED** [review]

File: framework/validators/IpValidator.php (hunk 3 above)
Comment (SilverFire) [reply framework/validators/IpValidator.php]: done

Comment (sergeymakinen) [reply framework/validators/IpValidator.php]: I think a deprecation is more than enough. BC break in this case is overkill.

Comment (SilverFire) [reply framework/validators/IpValidator.php]: It's a small break and I don't think a lot of people relied on internal constants of IpValidator

Comment (sergeymakinen) [reply framework/validators/IpValidator.php]: I do agree about internals. But the same applies to many framework areas (for example, `yii\db`) and it's a dangerous road I would not step onto.
IMHO, if we make an important feature everyone asks for years or fixing a long story bug, it's okay. But when we add a new helper, there's no need to rush. :)

File: framework/helpers/BaseIpHelper.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+     * ```
+     *
+     * @see https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing
+     *
+     * @param string $ip the valid IPv4 or IPv6 address
+     * @param int|string $cidr the CIDR of $ip
+     * @param string $range subnet in CIDR format e.g. `10.0.0.0/8` or `2001:af::/64`
+     * @return bool whether IP address with $cidr is in range
+     */
+    public static function inRange($ip, $cidr, $range)
```
Comment (Renkas) [inline framework/helpers/BaseIpHelper.php]: Why have separate cidr parameter for IP but expect CIDR to be inline on range?
I would expect the same logic for both.

Or am I missing somethng obvious here?

Comment (SilverFire) [reply framework/helpers/BaseIpHelper.php]: Yeah, good point :) It was ok as internal API of IpValidator, but should be re-designed for public API

**Comment by SilverFire** [comment]
Ready to merge

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**